File: Guigna/Guigna/GAgent.m
#import "GAgent.h"

@implementation GAgent

@synthesize appDelegate;

- (instancetype)init {
    self = [super init];
    if (self) {
        _processID = -1;
    }
    return self;
}

- (NSString *)outputForCommand:(NSString *)command {
    NSTask *task = [[NSTask alloc] init];
    NSArray *tokens = [command split];
    command = tokens[0];
    NSMutableArray *args = [NSMutableArray array];
    if ([tokens count] > 1) {
        NSArray *components = [tokens subarrayWithRange: NSMakeRange(1, [tokens count]-1)];
        for (NSString *component in components){
            if ([component isEqualToString:@"\"\""])
                [args addObject:[NSString string]];
            else
                [args addObject:[component stringByReplacingOccurrencesOfString:@"__" withString:@" "]];
        }
    }
    [task setLaunchPath:command];
    [task setArguments:args];
    NSPipe *pipe = [NSPipe pipe];
    [task setStandardOutput:pipe];
    NSPipe *errorPipe = [NSPipe pipe];
    [task setStandardError:errorPipe];
    [task setStandardInput:[NSPipe pipe]]; // NSLog workaround
    [task launch];
    _processID = [task processIdentifier];
    NSData *data = [[pipe fileHandleForReading] readDataToEndOfFile];
    [task waitUntilExit];
    _processID = -1;
    // int status = [task terminationStatus]; // TODO
    NSString __autoreleasing *output = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    // Uncomment to debug:
    // NSData *errorData = [[errorPipe fileHandleForReading] readDataToEndOfFile];
    // NSString __autoreleasing *errorOutput = [[NSString alloc] initWithData:errorData encoding:NSUTF8StringEncoding];
    return output;
}

- (NSArray *)nodesForURL:(NSString *)url XPath:(NSString *)xpath {
    NSError *error = nil;
    NSMutableString *page = [NSMutableString stringWithContentsOfURL:[NSURL URLWithString:url] encoding:NSUTF8StringEncoding error:&error];
    if (page == nil)
        page = [NSMutableString stringWithContentsOfURL:[NSURL URLWithString:url] encoding:NSISOLatin1StringEncoding error:&error];
    NSData *data = [page dataUsingEncoding:NSUTF8StringEncoding];
    NSXMLDocument __autoreleasing *doc = [[NSXMLDocument alloc] initWithData:data options:NSXMLDocumentTidyHTML error:&error];
    NSArray *nodes = [[doc rootElement] nodesForXPath:xpath error:&error];
    return nodes;
}

@end

File: Guigna-Swift/Guigna/GAgent.swift
import Foundation

class GAgent: NSObject {
    
    var appDelegate: GAppDelegate?
    var processID: CInt?
    
    func output(command: String) -> String {
        
        var task = NSTask()
        let tokens = command.componentsSeparatedByString(" ")
        let command = tokens[0]
        var args: [String] = []
        if tokens.count > 1 {
            let components = tokens[1...tokens.count-1]
            for component: String in components {
                if component == "\"\"" {
                    args.append("")
                } else {
                    args.append(component.stringByReplacingOccurrencesOfString("__", withString: " "))
                }
            }
        }
        task.launchPath = command
        task.arguments = args
        let pipe = NSPipe()
        task.standardOutput = pipe
        let errorPipe = NSPipe()
        task.standardError = errorPipe
        task.standardInput = NSPipe()
        task.launch()
        processID = task.processIdentifier
        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        task.waitUntilExit()
        processID = nil
        // int status = [task terminationStatus]; // TODO
        let output = NSString(data: data, encoding: NSUTF8StringEncoding)
        // Uncomment to debug:
        // NSData *errorData = [[errorPipe fileHandleForReading] readDataToEndOfFile];
        // NSString __autoreleasing *errorOutput = [[NSString alloc] initWithData:errorData encoding:NSUTF8StringEncoding];
        return output
    }
    
    
    func nodes(URL url: String, XPath xpath: String) -> [NSXMLNode] {
        // FIXME: doesn't allow xpath on childnodes
        var error: NSError? = nil
        var page: NSMutableString? = NSMutableString(contentsOfURL: NSURL(string: url), encoding: NSUTF8StringEncoding, error: &error)
        if page == nil {
            page = NSMutableString(contentsOfURL: NSURL(string: url), encoding: NSISOLatin1StringEncoding, error: &error)
        }
        var data: NSData = page!.dataUsingEncoding(NSUTF8StringEncoding)
        var doc = NSXMLDocument(data: data, options: Int(NSXMLDocumentTidyHTML), error: &error)
        var nodes = doc.rootElement().nodesForXPath(xpath, error: &error) as [NSXMLNode]
        return nodes
    }
}
